File: api/app/models/small_models.py
from tensorflow.keras import layers
from tensorflow.keras import Sequential
from tensorflow.keras.models import Model


def get_LSTM_v1() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.LSTM(38, return_sequences=True),
        layers.Dropout(0.1),
        layers.LSTM(16),
        layers.Dropout(0.2),
        layers.Dense(8, activation='relu'),
        layers.Dropout(0.3),
        layers.Dense(1, activation='sigmoid')
    ])


def get_LSTM_v2() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.LSTM(36, return_sequences=True),
        layers.Dropout(0.1),
        layers.LSTM(16),
        layers.Dropout(0.2),
        layers.Dense(8, activation='relu'),
        layers.Dropout(0.3),
        layers.Dense(1, activation='sigmoid')
    ])


def get_LSTM_v3() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.LSTM(36, return_sequences=True),
        layers.Dropout(0.1),
        layers.LSTM(16),
        layers.Dropout(0.3),
        layers.Dense(8, activation='relu'),
        layers.Dropout(0.4),
        layers.Dense(1, activation='sigmoid')
    ])


def get_LSTM_v4() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.LSTM(20, return_sequences=True),
        layers.Dropout(0.1),
        layers.LSTM(30),
        layers.Dropout(0.2),
        layers.Dense(10, activation='relu'),
        layers.Dropout(0.4),
        layers.Dense(1, activation='sigmoid')
    ])


def get_LSTM_v5() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.LSTM(38, return_sequences=True),
        layers.Dropout(0.1),
        layers.LSTM(16),
        layers.Dropout(0.2),
        layers.Dense(10, activation='relu'),
        layers.Dropout(0.4),
        layers.Dense(1, activation='sigmoid')
    ])


def get_GRU_v1() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.GRU(16, return_sequences=True),
        layers.Dropout(0.4),
        layers.GRU(32),
        layers.Dropout(0.4),
        layers.Dense(4, activation='relu'),
        layers.Dropout(0.2),
        layers.Dense(1, activation='sigmoid')
    ])


def get_GRU_v2() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.GRU(32, return_sequences=True),
        layers.Dropout(0.2),
        layers.GRU(8),
        layers.Dropout(0.3),
        layers.Dense(32, activation='relu'),
        layers.Dropout(0.2),
        layers.Dense(1, activation='sigmoid')
    ])


def get_GRU_v3() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.GRU(32, return_sequences=True),
        layers.Dropout(0.4),
        layers.GRU(8),
        layers.Dropout(0.4),
        layers.Dense(16, activation='relu'),
        layers.Dropout(0.3),
        layers.Dense(1, activation='sigmoid')
    ])


def get_GRU_v4() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.GRU(32, return_sequences=True),
        layers.Dropout(0.2),
        layers.GRU(8),
        layers.Dropout(0.2),
        layers.Dense(8, activation='relu'),
        layers.Dropout(0.5),
        layers.Dense(1, activation='sigmoid')
    ])


def get_GRU_v5() -> Model:
    return Sequential([
        layers.Input(shape=(60, 8)),
        layers.GRU(16, return_sequences=True),
        layers.Dropout(0.1),
        layers.GRU(12),
        layers.Dropout(0.5),
        layers.Dense(16, activation='relu'),
        layers.Dropout(0.2),
        layers.Dense(1, activation='sigmoid')
    ])


def get_model(type: str) -> Model:
    model = None

    match type:
        case 'L1':
            model = get_LSTM_v1()
        case 'L2':
            model = get_LSTM_v2()
        case 'L3':
            model = get_LSTM_v3()
        case 'L4':
            model = get_LSTM_v4()
        case 'L5':
            model = get_LSTM_v5()
        case 'G1':
            model = get_GRU_v1()
        case 'G2':
            model = get_GRU_v2()
        case 'G3':
            model = get_GRU_v3()
        case 'G4':
            model = get_GRU_v4()
        case 'G5':
            model = get_GRU_v5()

    if model:
        model.compile(
            optimizer='adam',
            loss='binary_crossentropy',
            metrics=['accuracy']
        )

    return model

File: worker/src/models/small-models.js
import { train, layers, sequential } from "@tensorflow/tfjs";
import { InputConfig, Initializers, InputShape } from "./utils.js";

export const SMALL_MODELS_POOL = [
  "G1",
  "L2",
  "G2",
  "D1",
  "G5",
  "G3",
  "G4",
  "D2",
  "D3",
];

const ModelsConfig = {
  L1: { l1: 24, d1: 0.4, l2: 24, d2: 0.2, l3: 14, d3: 0.1 },
  L2: { l1: 20, d1: 0.1, l2: 30, d2: 0.2, l3: 10, d3: 0.4 },
  L3: { l1: 30, d1: 0.4, l2: 20, d2: 0.2, l3: 18, d3: 0.1 },
  // ----------------------------------------------------------
  G2: { l1: 16, d1: 0.3, l2: 24, d2: 0.1, l3: 20, d3: 0.2 },
  G5: { l1: 10, d1: 0.4, l2: 24, d2: 0.1, l3: 20, d3: 0.2 },
  G3: { l1: 14, d1: 0.3, l2: 24, d2: 0.1, l3: 20, d3: 0.2 },
  G4: { l1: 14, d1: 0.3, l2: 24, d2: 0.1, l3: 20, d3: 0.3 },
  G1: { l1: 24, d1: 0.3, l2: 18, d2: 0.3, l3: 24, d3: 0.2 },
  G6: { l1: 32, d1: 0.2, l2: 8, d2: 0.3, l3: 32, d3: 0.2 },
  // ----------------------------------------------------------
  D2: { l1: 30, d1: 0.4, l2: 16, d2: 0.3, l3: 22, d3: 0.1 },
  D1: { l1: 24, d1: 0.2, l2: 20, d2: 0.3, l3: 20, d3: 0.1 },
  D3: { l1: 30, d1: 0.4, l2: 24, d2: 0.3, l3: 16, d3: 0.1 },
};

const get_LSTM = (config) => {
  return sequential({
    layers: [
      layers.lstm({ units: config.l1, ...InputConfig, ...Initializers }),
      layers.dropout({ rate: config.d1 }),
      layers.lstm({ units: config.l2, ...Initializers }),
      layers.dropout({ rate: config.d2 }),
      layers.dense({ units: config.l3, activation: "relu" }),
      layers.dropout({ rate: config.d3 }),
    ],
  });
};

const get_GRU = (config) => {
  return sequential({
    layers: [
      layers.gru({ units: config.l1, ...InputConfig, ...Initializers }),
      layers.dropout({ rate: config.d1 }),
      layers.gru({ units: config.l2, ...Initializers }),
      layers.dropout({ rate: config.d2 }),
      layers.dense({ units: config.l3, activation: "relu" }),
      layers.dropout({ rate: config.d3 }),
    ],
  });
};

const get_Dense = (config) => {
  return sequential({
    layers: [
      layers.flatten({ inputShape: InputShape }),
      layers.dense({ units: config.l1, activation: "relu" }),
      layers.dropout({ rate: config.d1 }),
      layers.dense({ units: config.l2, activation: "relu" }),
      layers.dropout({ rate: config.d2 }),
      layers.dense({ units: config.l3, activation: "relu" }),
      layers.dropout({ rate: config.d3 }),
    ],
  });
};

export const getModel = async (type) => {
  let model = undefined;

  switch (type[0]) {
    case "L":
      model = get_LSTM(ModelsConfig[type]);
      break;

    case "G":
      model = get_GRU(ModelsConfig[type]);
      break;

    case "D":
      model = get_Dense(ModelsConfig[type]);
      break;
  }

  model.add(layers.dense({ units: 1, activation: "sigmoid" }));

  const initLR = 0.001;
  const initMomentum = 0.9;
  const optimizer = train.momentum(initLR, initMomentum);

  model.compile({
    optimizer,
    loss: "binaryCrossentropy",
    metrics: ["accuracy"],
  });

  return model;
};
